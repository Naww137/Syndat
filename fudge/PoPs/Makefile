# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

SUB_DIRS = quantities chemicalElements families decays atomic Test

.PHONY: clean realclean doTarget

default:

clean:
	rm -f *.pyc
	$(MAKE) doTarget MODE=clean

realclean: clean
	$(MAKE) doTarget MODE=realclean

doTarget:
	SAVED_PWD=`pwd`; \
	for directory in $(SUB_DIRS); do cd $$directory; $(MAKE) $(MODE); cd $$SAVED_PWD; done
