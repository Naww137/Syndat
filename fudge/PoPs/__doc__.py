# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This is the 'Properties of Particles' or PoPs package. It defines a set of Python classes
for storing intrinsic particle properties like charge, mass, spin, etc.
"""
