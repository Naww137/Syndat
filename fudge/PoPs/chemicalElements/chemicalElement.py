# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the class for the suite chemicalElement.
"""

from .. import misc as miscModule

from ..atomic import atomic as atomicDataModule

from . import misc as chemicalElementMiscModule
from . import isotope as isotopeModule

class ChemicalElement( miscModule.ClassWithSymbolKey ) :
    """
    This class represents a chemical element (i.e., Oyxgen, Iron, Uranium). A chemical element
    stores all isotopes that contain the same number of protons (i.e., O16, O17, and O18). 
    The members of a chemicalElement are:

    .. table:: Chemical element members
       :widths: auto

       =========== =========== =========================================================================
       member      type        description
       =========== =========== =========================================================================
       symbol      string      This is the symbol for the chemical element (e.g., 'O' for oxygen).
       Z           int         Number of protons for this chemical element (e.g., 8 for 'O').
       name        string      The name of the chemical element (i.e., 'Oxygen')
       atomicData  atomicData  Container for atomic data, including ionized states and atomic relaxation
       isotopes    suite       The list of isotopes
       =========== =========== =========================================================================

    """

    moniker = 'chemicalElement'

    def __init__( self, symbol, Z, name ) :
        """
        :param symbol: Chemical symbol for the element, e.g. 'Fe' (string)
        :param Z: Atomic number, e.g. 26 (int)
        :param name: Full name of the element, e.g. 'Iron'
        """

        miscModule.ClassWithSymbolKey.__init__( self, symbol )

        base, anti = miscModule.baseAntiFromID( symbol )

        self.__anti = anti == miscModule.antiSuffix

        chemicalElementMiscModule.checkZ( Z )
        self.__Z = Z
        if( chemicalElementMiscModule.symbolFromZ[Z] != base ) : raise ValueError( 'Z = %s and chemical element symbol = "%s" are not consistent' %
                ( Z, miscModule.toLimitedString( symbol ) ) )

        if( not( isinstance( name, str ) ) ) : TypeError( 'name must be a string' )
        if( chemicalElementMiscModule.nameFromZ[Z] != name ) : 
            raise ValueError( 'Z = "%s" and chemical element name = "%s" are not consistent, should be "%s".' %
                    ( Z, miscModule.toLimitedString( name ), chemicalElementMiscModule.nameFromZ[Z] ) )
        self.__name = name

        self.__atomicData = None

        self.__isotopes = isotopeModule.Isotopes( )
        self.__isotopes.setAncestor( self ) 

    def __contains__( self, key ) : 

        return( key in self.__isotopes )

    def __getitem__( self, key ) :

        return( self.__isotopes[key] )

    def __iter__( self ) :
        """ Iterates over isotopes of this element (but not over the excited states of those isotopes) """

        for isotope in self.__isotopes : yield isotope

    @property
    def Z( self ) :

        return( self.__Z )

    @property
    def isAnti( self ) :

        return( self.__anti )

    @property
    def name( self ) :

        return( self.__name )

    @property
    def atomicData( self ) :

        return( self.__atomicData )

    @atomicData.setter
    def atomicData( self, value ) :

        if( not( isinstance( value, atomicDataModule.Atomic ) ) ) : raise TypeError( 'value must be an instance of atomicData.' )
        self.__atomicData = value
        self.__atomicData.setAncestor( self )

    @property
    def isotopes( self ) :

        return( self.__isotopes )

    def add( self, item ) :
        """
        :param item: isotope instance, will be added to the list of isotopes
        """

        self.__isotopes.add( item )

    def convertUnits( self, unitMap ) :
        """ See convertUnits documentation in PoPs.database """

        if( self.__atomicData is not None ) : self.__atomicData.convertUnits( unitMap )
        self.__isotopes.convertUnits( unitMap )

    def copy( self ) :
        """
        :return: deep copy of self
        """

        _chemicalElement = ChemicalElement( self.symbol, self.Z, self.name )
        if( self.__atomicData is not None ) : _chemicalElement.atomicData = self.__atomicData.copy( )
        for item in self.__isotopes : _chemicalElement.add( item.copy( ) )
        return( _chemicalElement )

    def sortCompare( self, other ) :

        if( not( isinstance( other, ChemicalElement ) ) ) : raise TypeError( 'Invalid other.' )
        return( self.Z - other.Z )

    def check(self, info):
        """ See check documentation in PoPs.database """

        from .. import warning as warningModule
        warnings = []

        for isotope in self:
            isotopeWarnings = isotope.check(info)
            if isotopeWarnings:
                warnings.append(warningModule.Context('Isotope %s' % isotope.symbol, isotopeWarnings))

        return warnings

    def toXML_strList( self, indent = '', **kwargs ) :

        indent2 = indent + kwargs.get( 'incrementalIndent', '  ' )

        XMLStringList = [ '%s<%s symbol="%s" Z="%s" name="%s">' % ( indent, self.moniker, self.symbol, self.Z, self.name ) ]
        if( self.__atomicData is not None ) : XMLStringList += self.__atomicData.toXML_strList( indent2, **kwargs )
        XMLStringList += self.__isotopes.toXML_strList( indent2, **kwargs )
        XMLStringList[-1] += '</%s>' % self.moniker
        return( XMLStringList )

    def parseNode(self, element, xPath, linkData, **kwargs):

        xPath.append( '%s[@symbol="%s"]' % ( element.tag, element.get( 'symbol' ) ) )

        for child in element :
            if( child.tag == isotopeModule.Isotopes.moniker ) :
                self.__isotopes.parseNode(child, xPath, linkData, **kwargs)
            elif( child.tag == atomicDataModule.Atomic.moniker ) :
                self.atomicData = atomicDataModule.Atomic.parseNodeUsingClass(child, xPath, linkData, **kwargs)
            else :
                raise ValueError( 'Invalid child = "%s" for %s' % ( child.tag, self.moniker ) )

        xPath.pop( )
        return( self )

    @classmethod
    def parseNodeUsingClass(cls, element, xPath, linkData, **kwargs):

        xPath.append( '%s[@symbol="%s"]' % ( element.tag, element.get( 'symbol' ) ) )

        self = cls( element.get('symbol'), int( element.get('Z') ), element.get('name') )
        xPath.pop( )

        self.parseNode(element, xPath, linkData, **kwargs)

        return( self )
