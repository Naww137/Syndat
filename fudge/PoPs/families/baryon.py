# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the baryon class definitions.
Baryons include the neutron and proton.
"""

from . import particle as particleModule

class Alias( particleModule.Alias ) :

    moniker = 'baryonAlias'

class Particle( particleModule.Particle ) :

    moniker = 'baryon'
    familyOrder = 2
    alias = Alias

    def __lt__( self, other ) :

        if( self.familyOrderLessThan( other ) ) : return( True )        # Also checks isinstance of other.
        if( self.familyOrder != other.familyOrder ) : return( False )
        return( self.getMass( 'amu' ) < other.getMass( 'amu' ) )

class Suite( particleModule.Suite ) :

    moniker = 'baryons'
    particle = Particle
