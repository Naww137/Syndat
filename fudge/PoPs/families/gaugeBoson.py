# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the gaugeBoson class definitions.
Gauge bosons include the photon, gluon, W and Z particles
"""

from . import particle as particleModule

class Alias( particleModule.Alias ) :

    moniker = 'gaugeBosonAlias'

class Particle( particleModule.Particle ) :

    moniker = 'gaugeBoson'
    familyOrder = 0
    alias = Alias

    def __lt__( self, other ) :

        if( self.familyOrderLessThan( other ) ) : return( True )        # Also checks isinstance of other.
        if( self.familyOrder != other.familyOrder ) : return( False )
        return( False )                                                 # Currently, only supports photon.

class Suite( particleModule.Suite ) :

    moniker = 'gaugeBosons'
    particle = Particle
