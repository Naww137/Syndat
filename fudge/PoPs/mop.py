# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

__doc__ = """
This module contains classes to handle a MoP (Map of PoPs) instance.
"""

import os
import re

from LUPY import ancestry as ancestryModule
from . import database as databaseModule

class FormatVersion :

    version_0_1 = '0.1'

    allowed = ( version_0_1, )

class Mop(ancestryModule.AncestryIO):

    moniker = 'mop'

    def __init__(self, name, path = './', format=FormatVersion.version_0_1):

        """
        Constructor for a Mop instance.

        :param name:            The name of the library.
        :param path:            The path to the mop file if read in from disk.
        :param format:          The format version for the mop.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( name, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__name = name 

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Library must be a string.' )
        self.__path = path

        if( format not in FormatVersion.allowed ) : raise ValueError( 'Invalid format "%s."' % format )
        self.__format = format

        self.__entries = []
        self.__pops = databaseModule.Database( 'Internal', '0.1' )

    def __len__( self ) :
        """Returns the number of entries in self."""

        return( len( self__entries ) )

    def __getitem__( self, index ) :
        """
        Returns the (index-1)^th item of self.

        :param index:       The index of the iter to return.
        """

        return( self.__entries[index] )

    @property
    def format( self ) :
        """Returns the format self is stored as."""

        return( self.__format )

    @property
    def name( self ) :
        """Returns self's name."""

        return( self.__name )

    @property
    def path( self ) :
        """Returns the path to self. If the data in self was read from disk, this is the path to the file."""

        return( self.__path )

    @property
    def fileName( self ) :
        """Same the path method."""

        return( self.__path )

    def append( self, entry ) :
        """Appends *entry* to the end of self's list of entries."""

        if( not( isinstance( entry, EntryBase ) ) ) : raise TypeError( 'Invalid instance to add.' )

        self.__entries.append( entry )
        entry.setAncestor( self )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns a list of items. Each item contains information about a PoPs database where one or more matches to *re_pid* exists. 
        Each item contains the list [ ids, all_ids, fileName, pops ]) where:

            -) ids:         The list of PoPs ids in a PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The file path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]).  *re_pid* can be a str instance or a *re.compile* instance.
        Note, even if *readPoPs* is False, the returned pops will not be None if it had previously been read in.
        
        :param re_pid:          A str or a *re.compile* pattern representing the ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        matches = []
        for entry in self.__entries : matches += entry.fetch( re_pid, readPoPs )

        return( matches )

    def toXML_strList( self, indent = "", **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        indent2 = indent + kwargs.get( 'incrementalIndent', '  ' )

        format = kwargs.get( 'format', FormatVersion.version_0_1 )
        if( format not in FormatVersion.allowed ) : raise TypeError( 'Invalid format = "%s".' % format )

        XML_list = [ '%s<%s name="%s" format="%s">' % ( indent, self.moniker, self.name, format ) ]
        for entry in self.__entries : XML_list += entry.toXML_strList( indent2, **kwargs )
        XML_list[-1] +=  '</%s>' % self.moniker

        return( XML_list )

    @staticmethod
    def parseNode( node, path ) :
        """
        Creates a Mop instance from an XML mop element.

        :param node:        XML element to parse.

        :return:            Mop instance.
        """

        if( node.tag != Mop.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Mop.moniker, node.tag ) )

        format = node.get( 'format', None )
        if( format is None ) : raise ValueError( 'Mop node does not have a "format" attribute.' )

        name = node.get( 'name', None )
        if( name is None ) : raise ValueError( 'Mop node does not have "name" attribute.' )

        mop = Mop( name, path = path )

        settings = { 'format' : format }
        for child in node :
            if( child.tag == Import.moniker ) :
                mop.append( Import.parseNode( settings, child ) )
            elif( child.tag == PoPs.moniker ) :
                mop.append( PoPs.parseNode( settings, child ) )
            else :
                raise ValueError( 'Invalid child tag "%s" for mop file.' % child.tag )

        return( mop )

class EntryBase(ancestryModule.AncestryIO) :
    """
    Base class for mop entries.
    """

    def __init__( self, path ) :
        """
        Base constructor for mop entries.

        :param path:            Path to the file for the entry.
        """

        ancestryModule.AncestryIO.__init__(self)

        if( not( isinstance( path, str ) ) ) : raise TypeError( 'Path must be a string.' )
        self.__path = path

    @property
    def fileName( self ) :
        """Returns the absolute path to the file pointed to by the 'path' attribute."""

        filename = self.path
        if( filename != os.sep ) : filename = os.path.join( os.path.dirname( os.path.realpath( self.ancestor.fileName ) ), filename )
        return( filename )

    @property
    def path( self ) :
        """Returns the path of the mop file. This may be absolute or relative."""

        return( self.__path )

class PoPs(EntryBase):
    """
    Class for storing a GNDS mop's PoPs instance.
    """

    moniker = 'pops'

    def __init__( self, path, ids ) :
        """
        Constructor for a mop's PoPs instance.

        :param path:            Path to the GNDS PoPs.database file.
        :param ids:             The list of particles ids in the PoPs.database file.
        """

        EntryBase.__init__(self, path)
        self.__pops = None
        self.__ids = ids

    @property
    def ids( self ) :
        """Returns the list of ids."""

        return( self.__ids )

    @property
    def pops( self ) :
        """Returns a reference to the PoPs.database instance. If the PoPs file has not be read in (see method *readIfNeeded*), None is returned."""

        return( self.__pops )

    def fetch( self, re_pid, readPoPs ) :
        """
        Returns the list [ ids, all_ids, fileName, pops ] if matches to *re_pid* are found in *self* and an empty list otherwise. 

            -) ids:         The list of PoPs ids in the PoPs database that matches *re_pid*,
            -) all_ids:     The list of all PoPs ids in the PoPs database,
            -) fileName:    The path to the PoPs database,
            -) pops:        An instance of the PoPs database if *readPoPs* is *True* and *None* otherwise.

        The *ids* are a list of str ids (e.g., [ 'Pu234', 'Pu234e_1' ]). *re_pid* can be a str instance or a *re.compile* instance.
        
        :param re_pid:          A str or *re.compile* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element.
        """

        if( isinstance( re_pid, str ) ) : re_pid = re.compile( re_pid )
        matches = [ pid for pid in self.ids if re_pid.fullmatch( pid ) ]

        if( len( matches ) > 0 ) :
            matches = [ [ matches, self.ids, self.fileName, self.readIfNeeded( readPoPs ) ] ]

        return( matches )

    def readIfNeeded( self, readPoPs = True ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in by a prior call and *readPoPs* 
        is True (default). Also, returns the value of *self.__pops*. The returned value will only be *None* if *readPoPs* 
        is *False* and all prior calls to this method had "*readPoPs* = False".

        :param readPoPs:        If *True*, the pops file is read.
        :return:                The value if *self.__pops*.
        """

        if( ( self.__pops is None ) and readPoPs ) :
            self.__pops = databaseModule.read( self.fileName )
            self.__pops.setAncestor( self )

        return( self.__pops )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s">' % ( indent, self.moniker, self.path ) + ' '.join( self.__ids ) + '</%s>' % self.moniker ] )

    @classmethod
    def parseNodeUsingClass(settings, node):
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != PoPs.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( PoPs.moniker, node.tag ) )

        path = node.get( 'path', None )
        ids = node.text.split( )

        pops = PoPs( path, ids )

        return( pops )

class Import(EntryBase):
    """
    Class for storing a GNDS mop's Import instance.
    """

    moniker = 'import'

    def __init__( self, path ) :
        """
        Constructor for a mop's import instance.

        :param path:            Path to the GNDS mop file.
        """

        EntryBase.__init__(self, path)
        self.__mop = None

    @property
    def mop( self ) :
        """Returns a reference to self.__mop."""

        return( self.__mop )

    def fetch( self, re_pid, readPoPs ) :
        """
        See documentation for Mop.fetch.
        
        :param re_pid:          A str or *re* pattern representing ids to match.
        :param readPoPs:        If *True*, the pops file is read and returned as the fourth element of each returned item.
        """

        self.readIfNeeded( )
        return( self.__mop.fetch( re_pid, readPoPs ) )

    def readIfNeeded( self ) :
        """
        Reads in the PoPs.database referenced by the *path* member if it has not been read in.
        """

        if( self.__mop is None ) :
            self.__mop = Mop.readXML_file( self.fileName )
            self.__mop.setAncestor( self )

    def toXML_strList( self, indent = '', **kwargs ) :
        """
        Returns a list of str instances representing the XML lines of self.

        :param indent:          The amount of indentation for each line. Child nodes and text may be indented more.
        :param **kwargs:        A keyword list.

        :return:                List of str instances representing the XML lines of self.
        """

        return( [ '%s<%s path="%s"/>' % ( indent, self.moniker, self.path ) ] )

    @staticmethod
    def parseNode( settings, node ) :
        """
        Creates an Import instance from an XML import element.

        :param settings:        Information from parent map instances needed to parse the XML element.
        :param node:            XML element to parse.

        :return:                Import instance.
        """

        if( node.tag != Import.moniker ) : raise TypeError( 'Invalid node name. Expected "%s" got "%s".' % ( Import.moniker, node.tag ) )

        path = node.get( 'path', None )

        _import = Import( path )

        return( _import )
