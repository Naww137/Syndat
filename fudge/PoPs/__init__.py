# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
Top-level __init__ for the PoPs package.
"""

POPS_MAJORVERSION = 0
POPS_MINORVERSION = 1
POPS_PATCHLEVEL = 0
__version__ = '%s.%s.%s' % ( POPS_MAJORVERSION, POPS_MINORVERSION, POPS_PATCHLEVEL )

from .__doc__ import __doc__
