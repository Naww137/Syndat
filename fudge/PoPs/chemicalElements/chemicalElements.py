# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

"""
This module contains the class for the suite chemicalElements.
"""

from .. import suite as suiteModule

from ..chemicalElements import misc as chemicalElementMiscModule

from . import isotope as isotopeModule
from . import chemicalElement as chemicalElementModule

from ..families import nuclide as nuclideModule
from ..families import nucleus as nucleusModule

class ChemicalElements( suiteModule.SortedSuite ) :
    """
    This class stores a list of chemicalElement instances.
    """

    moniker = 'chemicalElements'

    def __init__( self ) :
        """
        Creates an empty suite of chemicalElements. Use the add() method to fill the suite.
        """

        suiteModule.SortedSuite.__init__( self, [ chemicalElementModule.ChemicalElement ], replace = True )

    def __contains__( self, key ) :
        """
        Recursively search for key inside self and all child particle groups.
        :param key: symbol or id of a particle or particle group
        :return: True if key found
        """

        baseID, chemicalElementSymbol, A, levelID, isNucleus, anti, qualifier = chemicalElementMiscModule.chemicalElementALevelIDsAndAnti( key )
        if( chemicalElementSymbol is None ) : return( False )
        if( suiteModule.SortedSuite.__contains__( self, chemicalElementSymbol ) ) :
            return( key in suiteModule.SortedSuite.__getitem__( self, chemicalElementSymbol ) )
        return( False )

    def __getitem__( self, key ) :
        """
        :param key: string or integer
        :return: chemicalElement, nuclide or nucleus.  Raises KeyError if key not found
        """
        try:
            baseID, chemicalElementSymbol, A, levelID, isNucleus, anti, qualifier = chemicalElementMiscModule.chemicalElementALevelIDsAndAnti( key )
        except:
            chemicalElementSymbol = None

        if( chemicalElementSymbol is not None ) :
            chemicalElement = suiteModule.SortedSuite.__getitem__( self, chemicalElementSymbol )
            if( baseID == chemicalElementSymbol ) : return( chemicalElement )
            return( chemicalElement[key] )

        return suiteModule.SortedSuite.__getitem__( self, key )

    def add( self, item ) :
        """
        Adds a particle or particle group to the proper location inside self
        :param item: instance of chemicalElement, isotope, nuclide or nucleus
        """

        if( isinstance( item, chemicalElementModule.ChemicalElement ) ) :
            suiteModule.SortedSuite.add( self, item )
        elif( isinstance( item, ( isotopeModule.Isotope, nuclideModule.Particle, nucleusModule.Particle ) ) ) :
            chemicalElementSymbol = item.chemicalElementSymbol
            if( self.hasSymbol( chemicalElementSymbol ) ) :
                suiteModule.SortedSuite.__getitem__( self, chemicalElementSymbol ).add( item )
            else :
                chemicalElement = chemicalElementModule.ChemicalElement( chemicalElementSymbol, item.Z, chemicalElementMiscModule.nameFromZ[item.Z] )
                suiteModule.SortedSuite.add( self, chemicalElement )
                chemicalElement.add( item )
        else :
            raise TypeError( "Particle not a nuclide or nucleus object." )

    def convertUnits( self, unitMap ) :
        """ See convertUnits documentation in PoPs.database """

        for chemicalElement in self : chemicalElement.convertUnits( unitMap )

    def getSymbol( self, key ) :
        """
        Get the chemicalElement with symbol == key, or raises KeyError if no such chemicalElement is found.
        :param key: chemicalElement symbol
        :return: chemicalElement instance (or raises KeyError)
        """

        return( suiteModule.SortedSuite.__getitem__( self, key ) )

    def hasSymbol( self, key ) :
        """
        Search for chemicalElement with symbol == key. Unlike __contains__, this method does not
        recursively search inside child particle groups
        :param key: chemicalElement symbol
        :return: True if key present
        """

        return( suiteModule.SortedSuite.__contains__( self, key ) )

    def parseNode(self, element, xPath, linkData, **kwargs):

        xPath.append( element.tag )
        for child in element :
            self.add(chemicalElementModule.ChemicalElement.parseNodeUsingClass(child, xPath, linkData, **kwargs))
        xPath.pop()

    @classmethod
    def parseNodeUsingClass(cls, element, xPath, linkData, **kwargs):

        xPath.append( element.tag )
        self = cls( )
        xPath.pop()
        self.parseNode(element, xPath, linkData, **kwargs)

        return( self )
