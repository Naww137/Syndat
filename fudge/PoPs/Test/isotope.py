#! /usr/bin/env python3

# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

from pqu import PQU as PQUModule

from PoPs.chemicalElements import misc as chemicalElementMiscModule

from PoPs.quantities import quantity as quantityModule
from PoPs.quantities import mass as massModule
from PoPs.quantities import spin as spinModule
from PoPs.quantities import parity as parityModule
from PoPs.quantities import charge as chargeModule
from PoPs.quantities import halflife as halflifeModule
from PoPs.quantities import nuclearEnergyLevel as nuclearEnergyLevelModule

from PoPs.families import nuclide as nuclideModule 
from PoPs.chemicalElements import isotope as isotopeModule

A = 16
isotopeID = chemicalElementMiscModule.isotopeSymbolFromChemicalElementIDAndA( 'O', A )
isotope = isotopeModule.Isotope( isotopeID, A )

data = [ [ 0, 15.99491461956, 0,       None, None, None, None ],
         [ 1,           None, 6049400, None, None, None, None ],
         [ 2,           None, 6129893, None, None, None, None ],
         [ 3,           None, 6917100, None, None, None, None ] ]

for index, mass, energy, charge, halflife, spin, parity in data :
    name = chemicalElementMiscModule.nuclideIDFromIsotopeSymbolAndIndex( isotopeID, index )
    level = nuclideModule.Particle( name )
    energy = nuclearEnergyLevelModule.Double( 'base', energy, quantityModule.stringToPhysicalUnit( 'eV' ) )
    level.nucleus.energy.add( energy )

    if( mass is not None ) :
        mass = massModule.Double( 'base', mass, quantityModule.stringToPhysicalUnit( 'amu' ) )
        level.mass.add( mass )

    if( charge is not None ) :
        charge = chargeModule.Integer( 'base', charge, quantityModule.stringToPhysicalUnit( 'e' ) )
        level.charge.add( charge )

    if( halflife is not None ) :
        halflife = halflifeModule.Double( 'base', halflife, quantityModule.stringToPhysicalUnit( 's' ) )
        level.halflife.add( halflife )

    if( spin is not None ) :
        spin = spinModule.Fraction( 'base', spinModule.Fraction.toValueType( spin ), quantityModule.stringToPhysicalUnit( 'hbar' ) )
        level.spin.add( spin )

    if( parity is not None ) :
        parity = parityModule.Integer( 'base', parity, quantityModule.stringToPhysicalUnit( '' ) )
        level.parity.add( parity )

    isotope.add( level )

xmli1 = isotope.toXML( )
print( xmli1 )
isotope2 = isotopeModule.Isotope.parseXMLString(xmli1)
if( xmli1 != isotope2.toXML( ) ) : raise Exception( 'Fix me.' )

nuclide = isotope[isotopeID]
print( nuclide.mass.pqu( ) )
print( nuclide.mass.pqu( 'MeV/c**2' ) )
print( nuclide.mass.float( 'MeV/c**2' ) )

print()
print( nuclide.energy.pqu( ) )
print( nuclide.energy.pqu( 'MeV' ) )
print( PQUModule.floatToShortestString( nuclide.energy.float( 'MeV' ), 12 ) )
