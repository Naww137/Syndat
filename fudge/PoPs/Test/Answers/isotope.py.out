<isotope symbol="O16" A="16">
  <nuclides>
    <nuclide id="O16">
      <mass>
        <double label="base" value="15.9949146196" unit="amu"/></mass>
      <nucleus id="o16" index="0">
        <energy>
          <double label="base" value="0." unit="eV"/></energy></nucleus></nuclide>
    <nuclide id="O16_e1">
      <nucleus id="o16_e1" index="1">
        <energy>
          <double label="base" value="6.0494e6" unit="eV"/></energy></nucleus></nuclide>
    <nuclide id="O16_e2">
      <nucleus id="o16_e2" index="2">
        <energy>
          <double label="base" value="6129893." unit="eV"/></energy></nucleus></nuclide>
    <nuclide id="O16_e3">
      <nucleus id="o16_e3" index="3">
        <energy>
          <double label="base" value="6.9171e6" unit="eV"/></energy></nucleus></nuclide></nuclides></isotope>
15.99491461956 amu
14899.16852086884 MeV/c**2
14899.168520868838

0. eV
0. MeV
0
